File: java/src/ro/petrepatrasc/chainofresponsability/tests/bank/LoanSystemTest.java
package ro.petrepatrasc.chainofresponsability.tests.bank;

import com.tngtech.java.junit.dataprovider.DataProvider;
import com.tngtech.java.junit.dataprovider.DataProviderRunner;
import com.tngtech.java.junit.dataprovider.UseDataProvider;
import junit.framework.TestCase;
import org.junit.Test;
import org.junit.runner.RunWith;
import ro.petrepatrasc.chainofresponsability.agent.AbstractLoanAgent;
import ro.petrepatrasc.chainofresponsability.agent.BranchManager;
import ro.petrepatrasc.chainofresponsability.agent.LineManager;
import ro.petrepatrasc.chainofresponsability.agent.TillAgent;
import ro.petrepatrasc.chainofresponsability.bank.LoanSystem;
import ro.petrepatrasc.chainofresponsability.bank.LoanSystemBuilder;
import ro.petrepatrasc.chainofresponsability.customer.BankCustomer;
import ro.petrepatrasc.chainofresponsability.loan.LoanApplicationRequest;
import ro.petrepatrasc.chainofresponsability.loan.LoanApplicationResponse;

@RunWith(DataProviderRunner.class)
public class LoanSystemTest extends TestCase {

    @Test
    public void testThatTheFirstElementInTheChainHasBeenDefinedOnConstruction() {
        LoanSystem loanSystem = LoanSystemBuilder.build();
        AbstractLoanAgent firstInChain = loanSystem.getFirstInChain();

        this.assertNotNull(firstInChain);
    }

    @Test
    public void testThatAChainOfMoreThanOneAgentHasBeenDefinedOnConstruction() {
        LoanSystem loanSystem = LoanSystemBuilder.build();
        AbstractLoanAgent firstInChain = loanSystem.getFirstInChain();

        int membersInChain = 1;
        AbstractLoanAgent currentMember = firstInChain;

        while (null != currentMember.getNextInChain()) {
            membersInChain++;
            currentMember = currentMember.getNextInChain();
        }

        this.assertTrue(membersInChain > 1);
    }

    @Test
    @UseDataProvider("loanAmountDataProvider")
    public void testResolveLoanRequest(double loanAmount, boolean expectedAcceptance, Class agentClass) {
        LoanSystem loanSystem = LoanSystemBuilder.build();
        LoanApplicationRequest loanRequest = new LoanApplicationRequest(new BankCustomer("Unit Test Customer"), loanAmount);
        LoanApplicationResponse loanResponse = loanSystem.resolveLoanRequest(loanRequest);

        this.assertNotNull(loanResponse);
        this.assertEquals(expectedAcceptance, loanResponse.isApproved());
        this.assertTrue(loanResponse.getAgent().getClass().equals(agentClass));
    }

    @DataProvider
    public static Object[][] loanAmountDataProvider() {
        return new Object[][]{
                {312.17, true, TillAgent.class},
                {10.38, true, TillAgent.class},
                {512.75, true, TillAgent.class},
                {3985.48, true, TillAgent.class},
                {4999.99, true, TillAgent.class},
                {5120.75, true, BranchManager.class},
                {3985.48, true, TillAgent.class},
                {9999.99, true, BranchManager.class},
                {18360.28, true, LineManager.class},
                {15120.75, true, LineManager.class},
                {13985.48, true, LineManager.class},
                {19999.99, true, LineManager.class},
                {20000, false, LineManager.class},
                {25000.127, false, LineManager.class},
                {27158.89, false, LineManager.class},
        };
    }
}

File: php7/src/ChainOfResponsibility/Tests/Bank/LoanSystemTest.php
<?php

declare(strict_types = 1);

namespace PetrePatrasc\ChainOfResponsibility\Tests\Bank;


use PetrePatrasc\ChainOfResponsibility\Bank\LoanSystem;
use PetrePatrasc\ChainOfResponsibility\Bank\LoanSystemBuilder;
use PetrePatrasc\ChainOfResponsibility\BankCustomer;
use PetrePatrasc\ChainOfResponsibility\LoanApplicationRequest;

class LoanSystemTest extends \PHPUnit_Framework_TestCase
{
    /**
     * @var LoanSystem
     */
    protected $loanSystem;

    /**
     * @inheritDoc
     */
    protected function setUp()
    {
        parent::setUp();

        $this->loanSystem = LoanSystemBuilder::build();
    }

    public function testThatTheFirstElementInTheChainHasBeenDefinedOnConstruction()
    {
        $firstInChain = $this->loanSystem->getFirstInChain();

        $this->assertNotNull($firstInChain, 'No agents have been defined in the chain');
        $this->assertInstanceOf('\PetrePatrasc\ChainOfResponsibility\Agent\AbstractLoanAgent', $firstInChain);
    }

    public function testThatAChainOfMoreThanOneAgentHasBeenDefinedOnConstruction()
    {
        $firstInChain = $this->loanSystem->getFirstInChain();

        $membersInChain = 1;
        $currentMember = $firstInChain;

        while (null !== $currentMember->getNextInChain()) {
            $this->assertInstanceOf('\PetrePatrasc\ChainOfResponsibility\Agent\AbstractLoanAgent', $currentMember->getNextInChain(), 'The next member in the chain is not a loan agent.');

            $membersInChain++;
            $currentMember = $currentMember->getNextInChain();
        }

        $this->assertGreaterThan(1, $membersInChain, 'The chain should consist of more than one loan agents.');
    }

    /**
     * @param float $loanAmount
     * @param bool $expectedAcceptance
     * @param string $expectedAgentInstance
     *
     * @dataProvider loanApplicationDataProvider
     */
    public function testResolveLoanRequest(float $loanAmount, bool $expectedAcceptance, string $expectedAgentInstance)
    {
        $loanRequest = new LoanApplicationRequest(new BankCustomer('Unit Test Customer'), $loanAmount);
        $loanApprovalResult = $this->loanSystem->resolveLoanRequest($loanRequest);

        $this->assertNotNull($loanApprovalResult);
        $this->assertInstanceOf('\PetrePatrasc\ChainOfResponsibility\LoanApplicationResponse', $loanApprovalResult);

        $this->assertInternalType('bool', $loanApprovalResult->isApproved());
        $this->assertEquals($expectedAcceptance, $loanApprovalResult->isApproved());

        $this->assertNotNull($loanApprovalResult->getAgent());
        $this->assertInstanceOf($expectedAgentInstance, $loanApprovalResult->getAgent());
    }

    public function loanApplicationDataProvider(): array
    {
        $tillAgent = '\PetrePatrasc\ChainOfResponsibility\Agent\TillAgent';
        $branchManager = '\PetrePatrasc\ChainOfResponsibility\Agent\BranchManager';
        $lineManager = '\PetrePatrasc\ChainOfResponsibility\Agent\LineManager';

        return [
            [32.20, true, $tillAgent],
            [320, true, $tillAgent],
            [4999.99, true, $tillAgent],
            [5000, true, $branchManager],
            [9000, true, $branchManager],
            [9999.99, true, $branchManager],
            [10000, true, $lineManager],
            [15000.807, true, $lineManager],
            [19999.99, true, $lineManager],
            [20000, false, $lineManager],
            [30000, false, $lineManager],
        ];
    }

    /**
     * @param mixed $loanAmount
     *
     * @expectedException TypeError
     *
     * @dataProvider typeErrorDataProvider
     */
    public function testResolveLoanRequestTypeErrorException($loanAmount)
    {
        $loanRequest = new LoanApplicationRequest(new BankCustomer('Unit Test Customer'), $loanAmount);

        $this->loanSystem->applyForLoan($loanRequest);
    }

    public function typeErrorDataProvider(): array
    {
        return [
            ['32'],
            [new \stdClass()],
            [true]
        ];
    }
}
